Svc/BufferRepeater/BufferRepeater.cpp

```diff
@@ -0,0 +1,81 @@
+//
+// \copyright
+// Copyright 2009-2015, by the California Institute of Technology.
+// ALL RIGHTS RESERVED.  United States Government Sponsorship
+// acknowledged.
+//
+// ======================================================================
+
+#include <FpConfig.hpp>
+#include <Svc/BufferRepeater/BufferRepeater.hpp>
+
+namespace Svc {
+
+// ----------------------------------------------------------------------
+// Construction, initialization, and destruction
+// ----------------------------------------------------------------------
+
+BufferRepeater ::BufferRepeater(const char* const compName)
+    : BufferRepeaterComponentBase(compName),
+      m_allocation_failure_response(BufferRepeater::NUM_BUFFER_REPEATER_FAILURE_OPTIONS) {}
+
+void BufferRepeater ::init(const NATIVE_INT_TYPE instance) {
+    BufferRepeaterComponentBase::init(instance);
+}
+
+BufferRepeater ::~BufferRepeater() {}
+
+void BufferRepeater ::configure(BufferRepeater::BufferRepeaterFailureOption allocation_failure_response) {
+    this->m_allocation_failure_response = allocation_failure_response;
+}
+
+bool BufferRepeater ::check_allocation(FwIndexType index,
+                                       const Fw::Buffer& new_allocation,
+                                       const Fw::Buffer& incoming_buffer) {
+    bool is_valid = (new_allocation.getData() != nullptr) && (new_allocation.getSize() >= incoming_buffer.getSize());
+
+    // Respond to invalid buffer allocation
+    if (!is_valid) {
+        switch (this->m_allocation_failure_response) {
+            case NO_RESPONSE_ON_OUT_OF_MEMORY:
+                // No response intended
+                break;
+            case WARNING_ON_OUT_OF_MEMORY:
+                this->log_WARNING_HI_AllocationSoftFailure(index, incoming_buffer.getSize());
+                break;
+            case FATAL_ON_OUT_OF_MEMORY:
+                this->log_FATAL_AllocationHardFailure(index, incoming_buffer.getSize());
+                break;
+            default:
+                FW_ASSERT(0);
+                break;
+        }
+    }
+    return is_valid;
+}
+
+// ----------------------------------------------------------------------
+// Handler implementations for user-defined serial input ports
+// ----------------------------------------------------------------------
+
+void BufferRepeater ::portIn_handler(NATIVE_INT_TYPE portNum, /*!< The port number*/
+                                     Fw::Buffer& buffer       /*!< The serialization buffer*/
+) {
+    FW_ASSERT(this->m_allocation_failure_response < NUM_BUFFER_REPEATER_FAILURE_OPTIONS);
+    for (FwIndexType i = 0; i < NUM_PORTOUT_OUTPUT_PORTS; i++) {
+        if (isConnected_portOut_OutputPort(i)) {
+            Fw::Buffer new_allocation = this->allocate_out(0, buffer.getSize());
+            if (this->check_allocation(i, new_allocation, buffer)) {
+                // Clone the data and send it
+                ::memcpy(new_allocation.getData(), buffer.getData(), buffer.getSize());
```

Comment:
Don't we have to set the destination buffer size to be the same as the source so users know how much data there is?

Reply:
Yes, yes we do!  Good catch.  Also need to put this in UTs.